crates/rethnet_evm/src/db/fork.rs

```diff
@@ -0,0 +1,271 @@
+                    .storage(address, index)
+                    .map_err(ForkDatabaseError::RemoteDatabase)?;
+
+                self.storage_cache.insert((address, index), remote);
+
+                Ok(remote)
+            }
+        }
+    }
+}
+
+impl revm::DatabaseCommit for ForkDatabase {
+    fn commit(&mut self, changes: HashMap<Address, Account>) {
+        self.layered_db.commit(changes)
+    }
+}
+
+impl crate::DatabaseDebug for ForkDatabase {
+    type Error = ForkDatabaseError;
+
+    fn account_storage_root(&mut self, address: &Address) -> Result<Option<B256>, Self::Error> {
+        self.layered_db
+            .account_storage_root(address)
+            .map_err(ForkDatabaseError::LayeredDatabase)
+    }
+
+    /// Inserts an account with the specified address.
+    fn insert_account(
+        &mut self,
+        address: Address,
+        account_info: AccountInfo,
+    ) -> Result<(), Self::Error> {
+        self.layered_db
+            .insert_account(address, account_info)
+            .map_err(ForkDatabaseError::LayeredDatabase)
+    }
+
+    /// Modifies the account at the specified address using the provided function.
+    fn modify_account(
+        &mut self,
+        address: Address,
+        modifier: crate::debug::ModifierFn,
+    ) -> Result<(), Self::Error> {
+        self.layered_db
+            .modify_account(address, modifier)
+            .map_err(ForkDatabaseError::LayeredDatabase)
+    }
+
+    /// Removes and returns the account at the specified address, if it exists.
+    fn remove_account(&mut self, address: Address) -> Result<Option<AccountInfo>, Self::Error> {
+        crate::DatabaseDebug::remove_account(&mut self.layered_db, address)
+            .map_err(ForkDatabaseError::LayeredDatabase)
+    }
+
+    /// Sets the storage slot at the specified address and index to the provided value.
+    fn set_account_storage_slot(
+        &mut self,
+        address: Address,
+        index: U256,
+        value: U256,
+    ) -> Result<(), Self::Error> {
+        self.layered_db
+            .set_account_storage_slot(address, index, value)
+            .map_err(ForkDatabaseError::LayeredDatabase)
+    }
+
+    /// Reverts the state to match the specified state root.
+    fn set_state_root(&mut self, state_root: &B256) -> Result<(), Self::Error> {
+        self.layered_db
+            .set_state_root(state_root)
+            .map_err(ForkDatabaseError::LayeredDatabase)
+    }
+
+    /// Retrieves the storage root of the database.
+    fn state_root(&mut self) -> Result<B256, Self::Error> {
+        if self.layered_db.iter().next().is_some() {
+            Ok(self
+                .layered_db
+                .state_root()
+                .map_err(ForkDatabaseError::LayeredDatabase)?)
+        } else if let Some(cached) = self.fork_block_state_root_cache {
+            Ok(cached)
+        } else {
+            self.fork_block_state_root_cache = Some(
+                self.remote_db
```

Comment:
We might have to create a workaround for this. `state_root`s in Hardhat forked databases are pseudorandomly generated hashes.

This struct mimics the Hardhat implementation: https://github.com/NomicFoundation/hardhat/blob/rethnet/main/crates/rethnet_evm/src/random.rs#L8

And this is the original TypeScript implementation that you can track to see how it's used in Hardhat: https://github.com/NomicFoundation/hardhat/blob/rethnet/main/packages/hardhat-core/src/internal/hardhat-network/provider/utils/random.ts

Reply:
Are you sure we need to do anything with that here in this PR? It seems to me that perhaps we'd want to use that within the `LayeredDatabase`, when new layers are created, but that when referring to historical blocks (fork block or prior) we'd want to just use the actual state root from the block.

Reply:
The logic should be independent from `LayeredDatabase` as we only need pseudorandom root hashes when using `ForkDatabase`. When just using `LayeredDatabase` we can always calculate our state root hashes.

When combining the `LayeredDatabase` with a remote database, it needs to somehow create a unique root hash for each layer that takes into account the "history" as well.

This is what the `ForkStateManager` does in Hardhat:

https://github.com/NomicFoundation/hardhat/blob/rethnet/main/packages/hardhat-core/src/internal/hardhat-network/provider/fork/ForkStateManager.ts#L272-L278